File: CSE-106 (Data Structure Lab)/lab-report/report-1/final/main.go
// Written by: Md. Maruf Sarker
// Platform: Arch Linux
// Problem: Binary Search (Use Bubble Sort to sort the elements)
// Language: go
// Date: 2022-10-26

package main

import (
	"fmt"
)

// Function to sort the array using Bubble Sort
func bubbleSort(arr []int, n int) {
	for i := 0; i < n-1; i++ {
		for j := 0; j < n-i-1; j++ {
			if arr[j] > arr[j+1] {
				arr[j], arr[j+1] = arr[j+1], arr[j]
			}
		}
	}
}

// Function to search the key in the array using Binary Search
func binarySearch(arr []int, n int, key int) {
	low := 0
	high := n - 1
	mid := 0
	for low <= high {
		mid = (low + high) / 2
		if arr[mid] == key {
			fmt.Println("Found at index:", mid)
			return
		} else if arr[mid] < key {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}
	fmt.Println("Not Found")
}

func main() {
	var n, i int
	var arr [100]int

	fmt.Print("Enter the number of elements: ")
	fmt.Scan(&n)

	fmt.Print("Enter the elements: ")
	for i = 0; i < n; i++ {
		fmt.Scan(&arr[i])
	}

	// Bubble Sort
	bubbleSort(arr[:], n)

	fmt.Print("Sorted array: ")
	for i = 0; i < n; i++ {
		fmt.Print(arr[i], " ")
	}

	// Binary Search
	var key int

	fmt.Print("\nEnter the key: ")
	fmt.Scan(&key)

	binarySearch(arr[:], n, key)
}

File: CSE-106 (Data Structure Lab)/lab-report/report-1/final/main.cpp
// Written by: Md. Maruf Sarker
// Platform: Arch Linux
// Problem: Binary Search (Use Bubble Sort to sort the elements)
// Language: cpp
// Date: 2022-10-26

#include <bits/stdc++.h>
using namespace std;

#define endl "\n"

// Function to sort the array using Bubble Sort
void bubbleSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

// Function to search the key in the array using Binary Search
int binarySearch(int arr[], int n, int key) {
    int low = 0, high = n - 1;
    while (low <= high) {
        int mid = (low + high) / 2;
        if (arr[mid] == key) {
            return mid;
        } else if (arr[mid] < key) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return -1;
}

int main() {
    // Variable Declaration and Initialization
    int n, i, j, temp, key, low, high, mid, flag = 0;

    // Input the number of elements
    cout << "Enter the number of elements: ";
    cin >> n;

    // Declare an array of size n
    int arr[n];
    cout << "Enter the elements: ";

    // Input the elements from the user
    for (i = 0; i < n; i++)
        cin >> arr[i];

    // Sort the elements using Bubble Sort
    bubbleSort(arr, n);

    // Sorted array
    cout << "Sorted array: ";
    for (i = 0; i < n; i++)
        cout << arr[i] << " ";

    // Input the key to be searched
    cout << "\nEnter the key to be searched: ";
    cin >> key;

    // Binary Search
    int pos = binarySearch(arr, n, key);

    // Print the result
    if (pos == -1) cout << "Key not found!" << endl;
    else cout << "Key found at position " << pos + 1 << endl;
    
    return 0;
}